dhall/src/Dhall/TypeCheck.hs

```diff
@@ -1308,11 +1316,14 @@ data TypeMessage s a
     | AlternativeAnnotationMismatch Text (Expr s a) Const Text (Expr s a) Const
     | ListAppendMismatch (Expr s a) (Expr s a)
     | MustCombineARecord Char (Expr s a) (Expr s a)
+    | InvalidDuplicateField Text (Expr s a) (Expr s a)
     | InvalidRecordCompletion Text (Expr s a)
     | CompletionSchemaMustBeARecord (Expr s a) (Expr s a)
     | CombineTypesRequiresRecordType (Expr s a) (Expr s a)
     | RecordTypeMismatch Const Const (Expr s a) (Expr s a)
-    | FieldCollision Text
+    | DuplicateFieldCannotBeMerged (NonEmpty Text)
+    | FieldCollision [Text]
+    | FieldTypeCollision [Text]
```

Comment:
It seems that we know that these "paths" are non-empty, so we should use `NonEmpty` instead of `[]`.